### PR TITLE
New demo seed data featuring scheduling mode

### PR DESCRIPTION
**Pull request by wielinde**
https://community.openproject.com/projects/openproject/work_packages/32882

### CODE REVIEW

**Comment by wielinde** [comment]
@lindenthal Could you please check the state of this PR so that we still can merge it before the 11.0 release?

- I adopted the onboarding tour to not mention features and bugs
- I created some very simple Task board with the tasks that were present in your data.

**Comment by lindenthal** [comment]
Hi @wielinde,

Thanks, I will have a look as soon as pullpreview is ready.

**Comment by wielinde** [comment]
@lindenthal Sorry, I didn't see your comments in the WP. Will work on that now.

File: app/seeders/demo_data/work_package_board_seeder.rb
Comment (ulferts) [inline app/seeders/demo_data/work_package_board_seeder.rb]: Given this recurring pattern:
```
print 'something'
  do_something
puts
```

it would make sense to introduce a method like this,
```
with_print 'something' do
  do_something
end
```

File: app/seeders/demo_data/work_package_board_seeder.rb
Comment (ulferts) [inline app/seeders/demo_data/work_package_board_seeder.rb:96]: This can be simplified to 

```
board.options[:filters] = filters_conf.select { |f| f[:type] }.map do |filter|
  type = Type.find_by(name: translate_with_base_url(filter[:type]))
  { type: { operator: '=', values: [type.id.to_s] } }
end
```

However, the better approach here would be to avoid hitting the database once per iteration and instantiating the whole Type. As the code's intend is to create a filter for every type id you could also write:

```
filter_names = filters_conf.select { |f| f[:type] }.map { |f| translate_with_base_url(filter[:type]) }
board.options[:filters] = Type.where(name: filter_names).pluck(:id).map do |id|
  { type: { operator: '=', values: [id.to_s] } }
end
```

Additionally, please avoid nesting this deeply, e.g. the initial condition `project_data_for(key, 'boards.kanban.filters').present?` wrapping the whole of the method body, could be turned into a guard statment:

```
filters_conf = project_data_for(key, 'boards.kanban.filters')

return if filters_conf.nil?
```

File: app/seeders/demo_data/work_package_board_seeder.rb
Comment (ulferts) [inline app/seeders/demo_data/work_package_board_seeder.rb:162]: Those should be I18n`ed.

While you are at it, could you please ensure, that throughout this file no reference strings are used, e.g. 'Scrum project' or 'Setup conference website'. Otherwise, it will break if ever the name of the project or work package is changed which I think rather likely to happen.

File: app/seeders/demo_data/work_package_board_seeder.rb
Comment (ulferts) [inline app/seeders/demo_data/work_package_board_seeder.rb:204]: Easy performance boost: 

```
subjects = [[I18n.t(...), I18n.t(...)], [I18n.t(...)] 

id_by_subjects = WorkPackage
  .where(subject: subjects.flatten)
  .pluck(:subject, :id)
  .to_h

subjects
  .map { |subselect| id_by_subjects.slice(*subselect).values }
  .concat([[]])
```  

Ideally, the mapping would be extracted into a method so that `basic_query_work_packages` as well as `scrum_query_work_packages` would benefit from it.

File: app/seeders/demo_data/work_package_board_seeder.rb
Comment (ulferts) [inline app/seeders/demo_data/work_package_board_seeder.rb:234]: Again, the method mapping subjects to ids could be employed.

**Review by ulferts — CHANGES_REQUESTED** [review]
@wielinde as agreed, I did not look into the resulting data but merely reviewed the code. Please find my comments inline.

File: app/seeders/demo_data/work_package_board_seeder.rb
Comment (ulferts) [reply app/seeders/demo_data/work_package_board_seeder.rb:234]: Forget about this one, I was mistaken.

File: app/seeders/demo_data/work_package_board_seeder.rb
Comment (wielinde) [reply app/seeders/demo_data/work_package_board_seeder.rb:162]: As discussed in person that redesign would be to much for this PR. 

However, we sketched a structure that sounds interesting to pursue in the future:

We should seperate seed data structure form their translation. So we would move everything I18n related to `en.yml` and reference those strings from within the structure giving YML file i.e. `seeders.standard.yml`. 

A bit tricky will be to still support the seed data extraction rake tasks. They would also separate into I18n strings and structure (i.e. for relations). We would need to introduce some kind of WP keys such as: `<hostname>-<wp-id>--<subject like slug>`. With such a pattern we could have some increased persistency. We would only check for hostname and wp-id and ignore the slug, as the subject could change accidentally.

File: app/seeders/demo_data/work_package_board_seeder.rb
Comment (wielinde) [reply app/seeders/demo_data/work_package_board_seeder.rb]: Agree.

File: app/seeders/demo_data/work_package_board_seeder.rb
Comment (wielinde) [reply app/seeders/demo_data/work_package_board_seeder.rb:96]: Yes, of cause, the guard statement is much nicer.

Regarding the suggested refactoring I believe we understood a different intentions here. My intention is less allowing filtering on multiple types at the same time. I rather intend to have some easy way to add other non-type filters at a later point, such as status etc. If it's ok with you, I'd leave it as it is and we can still optimize it for whatever situation we find ourselves in in the future.

**Comment by ulferts** [comment]
I am going ahead and simply merge it now. The functionality is fine. It is only the code that is lacking and this can still be approved after this merge. This will give us more time for other topics (e.g. i18n) that need to be addressed before the release.

**Pull request merged**
